File: src/main/java/com/_14ercooper/megaserver/UserInput.java
package com._14ercooper.megaserver;

import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class UserInput {
    // Get an input in a numeric range
    public static int numberRange(int min, int max) {
        try {
            int i = Integer.parseInt(System.console().readLine());
            if (i >= min && i <= max)
                return i;
            else {
                System.out.print("Input outside of valid range.\n> ");
                return numberRange(min, max);
            }
        } catch (Exception e) {
            System.out.print("Input invalid.\n  > ");
            return numberRange(min, max);
        }
    }

    public static String patternMatch(String regex) {
        try {
            String s = System.console().readLine();
            Pattern p = Pattern.compile(regex);
            Matcher m = p.matcher(s);
            if (m.find())
                return s;
            else {
                System.out.print("Input does not match required pattern.\n> ");
                return patternMatch(regex);
            }
        } catch (Exception e) {
            System.out.print("Input invalid.\n> ");
            return patternMatch(regex);
        }
    }

    public static String fromList(List<String> choices) {
        int i = 1;
        for (String s : choices) {
            System.out.println("\t" + i + ") " + s);
            i++;
        }
        System.out.print("> ");
        return choices.get(numberRange(1, i) - 1);
    }
}

File: src/main/java/com/_14ercooper/megaserver/UserInput.kt
/**
 * This file is part of 14erEdit.
 * 
  * 14erEdit is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * 14erEdit is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with 14erEdit.  If not, see <https://www.gnu.org/licenses/>.
 */

package com._14ercooper.megaserver

import java.util.regex.Pattern

object UserInput {
    // Get an input in a numeric range
    fun numberRange(min: Int, max: Int): Int {
        return try {
            val i = System.console().readLine().toInt()
            if (i in min..max) i else {
                print("Input outside of valid range.\n> ")
                numberRange(min, max)
            }
        } catch (e: Exception) {
            print("Input invalid.\n  > ")
            numberRange(min, max)
        }
    }

    fun patternMatch(regex: String): String {
        return try {
            val s = System.console().readLine()
            val p = Pattern.compile(regex)
            val m = p.matcher(s)
            if (m.find()) s else {
                print("Input does not match required pattern.\n> ")
                patternMatch(regex)
            }
        } catch (e: Exception) {
            print("Input invalid.\n> ")
            patternMatch(regex)
        }
    }

    fun fromList(choices: List<String>): String {
        var i = 1
        for (s in choices) {
            println("\t$i) $s")
            i++
        }
        print("> ")
        return choices[numberRange(1, i) - 1]
    }
}